src/core/mac/mac_frame.cpp

```diff
@@ -254,27 +253,107 @@ Error Frame::ValidatePsdu(void) const
     return error;
 }
 
+#if OPENTHREAD_CONFIG_WAKEUP_COORDINATOR_ENABLE || OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+bool Frame::IsWakeupFrame(void) const
+{
+    const uint16_t fcf    = GetFrameControlField();
+    bool           result = false;
+    uint8_t        keyIdMode;
+    uint8_t        firstIeIndex;
+    const uint8_t *rendezvousIe;
+    const uint8_t *connectionIe;
+    Address        srcAddress;
+
+    // Wake-up frame is a Key Id Mode 2 secured...
+    SuccessOrExit(GetKeyIdMode(keyIdMode));
+    VerifyOrExit(keyIdMode == kKeyIdMode2);
+
+    // Multipurpose frame without Ack Request...
+    VerifyOrExit((fcf & kFcfFrameTypeMask) == kTypeMultipurpose);
+    VerifyOrExit((fcf & kMpFcfAckRequest) == 0);
```

Comment:
Should we check these first before `KeyIdMode` as they are easier/quicker to check and we can exit more quickly?